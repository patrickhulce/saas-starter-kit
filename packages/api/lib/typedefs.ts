import 'express' // tslint:disable-line

declare module 'express-serve-static-core' {
  /* tslint:disable */
  export interface Response {
    body?: any
  }
  /* tslint:enable */
}

export {}
