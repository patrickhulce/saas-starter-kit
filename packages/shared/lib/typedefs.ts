import {ISQLOptions} from 'klay'

export enum ModelID {
  Account = 'account',
  User = 'user',
}

export interface IAccount {
  id?: number
  name: string
  plan: AccountPlan
  createdAt?: Date
  updatedAt?: Date
}

export interface IUser {
  id?: number
  accountId: number
  role: AuthRole
  email: string
  password: string
  firstName: string
  lastName: string
  isVerified: boolean
  verificationKey?: string
  createdAt?: Date
  updatedAt?: Date
}

export enum AuthRole {
  Root = 'root',
  Admin = 'admin',
  User = 'user',
  Anonymous = 'anonymous',
}

export enum Permission {
  RootAccess = '_root_',
  AccountManage = 'account:manage',
  AccountView = 'account:view',
  UserPassword = 'user:password',
  UserProfile = 'user:profile',
  UserView = 'user:view',
}

export enum AccountPlan {
  Free = 'free',
  Bronze = 'bronze',
  Silver = 'silver',
  Gold = 'gold',
}

export interface IConf {
  debug: boolean
  production: boolean
  isUnderTest: boolean

  origin: string
  cdnAppURL: string
  termsOfServiceURL: string
  displayName: string

  secret: string
  database: ISQLOptions

  mailslurp: {apiKey: string}
  sparkpost: {fromAddress: string; apiKey: string; sendToSink: boolean}
}
