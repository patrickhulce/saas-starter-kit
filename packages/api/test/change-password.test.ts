import {IState} from './typedefs'

describe('change password', () => {
  const state: IState = {}

  require('./steps/initialize.test')(state)
  require('./steps/account-setup.test')(state)

  describe('change password', () => {
    it('should reject unauthenticated password changes', async () => {
      const response = await fetch(`${state.apiURL}/v1/users/${state.user.id}/password`, {
        method: 'PUT',
        body: JSON.stringify({password: 'new-password1'}),
        headers: {'content-type': 'application/json'},
      })

      expect(response.status).toBe(401)
    })

    it('should reject password changes that do not confirm existing password', async () => {
      const response = await fetch(`${state.apiURL}/v1/users/${state.user.id}/password`, {
        method: 'PUT',
        body: JSON.stringify({password: 'new-password1'}),
        headers: {'content-type': 'application/json', cookie: `token=${state.token}`},
      })

      expect(response.status).toBe(400)
      expect(await response.json()).toMatchSnapshot()
    })

    it('should accept password changes', async () => {
      const response = await fetch(`${state.apiURL}/v1/users/${state.user.id}/password`, {
        method: 'PUT',
        body: JSON.stringify({password: 'new-password1'}),
        headers: {
          'content-type': 'application/json',
          cookie: `token=${state.token}`,
          'x-current-password': 'password1',
        },
      })

      expect(response.status).toBe(200)
    })

    it('should not login with old password', async () => {
      const payload = {
        grant_type: 'password',
        username: state.login.email,
        password: state.login.password,
      }

      const response = await fetch(`${state.apiURL}/v1/oauth/token`, {
        method: 'POST',
        body: JSON.stringify(payload),
        headers: {'content-type': 'application/json'},
      })

      expect(response.status).toBe(401)
    })

    it('should login with new password', async () => {
      const payload = {
        grant_type: 'password',
        username: state.login.email,
        password: 'new-password1',
      }

      const response = await fetch(`${state.apiURL}/v1/oauth/token`, {
        method: 'POST',
        body: JSON.stringify(payload),
        headers: {'content-type': 'application/json'},
      })

      expect(response.status).toBe(200)
      state.login.password = 'new-password1'
    })
  })

  require('./steps/teardown.test')(state)
})
