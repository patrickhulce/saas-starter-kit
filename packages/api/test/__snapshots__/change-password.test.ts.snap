// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change password change password should reject password changes that do not confirm existing password 1`] = `
Object {
  "message": "current password invalid",
  "name": "AssertionError",
}
`;
